L1Trigger/L1THGCal/src/HGCalTriggerTools.cc

```diff
@@ -44,23 +109,84 @@ layers(ForwardSubdetector type) const
 
 unsigned
 HGCalTriggerTools::
-layerWithOffset(unsigned id) const
-{
-  HGCalDetId detid(id);
-  unsigned layer = 0;
-  switch(detid.subdetId())
-  {
-    case ForwardSubdetector::HGCEE:
-      layer = detid.layer();
-      break;
-    case ForwardSubdetector::HGCHEF:
-      layer = eeLayers_ + detid.layer();
-      break;
-    case ForwardSubdetector::HGCHEB:
-      layer = eeLayers_ + fhLayers_ + detid.layer();
-      break;
-    default:
-      break;
-  };
+layer(const DetId& id) const {
+  unsigned int layer = std::numeric_limits<unsigned int>::max();
+  if( id.det() == DetId::Forward) {
+    const HGCalDetId hid(id);
+    layer = hid.layer();
+  } else if( id.det() == DetId::Hcal && id.subdetId() == HcalEndcap) {
+    const HcalDetId hcid(id);
+    layer = hcid.depth();
+  }
   return layer;
 }
+
+unsigned 
+HGCalTriggerTools::
+layerWithOffset(const DetId& id) const {
+  unsigned int l = layer(id);
+  if( id.det() == DetId::Forward && id.subdetId() == HGCHEF ) {
+    l += eeLayers_;
+  } else if( (id.det() == DetId::Hcal && id.subdetId() == HcalEndcap) ||
+             (id.det() == DetId::Forward && id.subdetId() == HGCHEB) ) {
+    l += eeLayers_ + fhLayers_;
+  }
+  return l;
+}
+
+float HGCalTriggerTools::getEta(const GlobalPoint& position, const float& vertex_z) const {
+  GlobalPoint corrected_position = GlobalPoint(position.x(), position.y(), position.z()-vertex_z);
+  return corrected_position.eta();
+}
+
+float HGCalTriggerTools::getTCEta(const DetId& id, const float& vertex_z) const {
+  GlobalPoint position = getTCPosition(id);
+  return getEta(position, vertex_z);
+}
+
+float HGCalTriggerTools::getPhi(const GlobalPoint& position) const {
+  float phi = atan2(position.y(),position.x());
+  return phi;
+}
+
+float HGCalTriggerTools::getTCPhi(const DetId& id) const {
+  GlobalPoint position = getTCPosition(id);
+  return getPhi(position);
+}
+
+float HGCalTriggerTools::getPt(const GlobalPoint& position, const float& hitEnergy, const float& vertex_z) const {
+  float eta = getEta(position, vertex_z);
+  float pt = hitEnergy / cosh(eta);
+  return pt;
+}
+
+float HGCalTriggerTools::getTCPt(const DetId& id, const float& hitEnergy, const float& vertex_z) const {
+  GlobalPoint position = getTCPosition(id);
+  return getPt(position, hitEnergy, vertex_z);
+}
+
+float HGCalTriggerTools::getLayerZ(const unsigned& layerWithOffset) const {
+  int subdet = ForwardSubdetector::HGCEE;
+  unsigned offset = 0;
+  if(layerWithOffset > lastLayerEE() && layerWithOffset <= lastLayerFH()) {
+    subdet = ForwardSubdetector::HGCHEF;
+    offset = lastLayerEE();
+  } else if(layerWithOffset > lastLayerFH()) {
+    subdet = HcalSubdetector::HcalEndcap;
+    offset = lastLayerFH();
+  }
+  return getLayerZ(subdet, layerWithOffset - offset);
+}
+
+float HGCalTriggerTools::getLayerZ(const int& subdet, const unsigned& layer) const {
+  const unsigned heOffset = 7;
```

Comment:
can this be obtained from geometry rather than hardcoded?

Reply:
This has been modified to rely more on the geometry, but I must say I don't really understand the logic in the `getRZ()` function of the HCAL DDDconstants class